js/repl/CodeMirror.tsx

```diff
@@ -1,138 +1,134 @@
-      this._codeMirror.getValue() !== nextProps.value
-    ) {
-      if (nextProps.preserveScrollPosition) {
-        const prevScrollPosition = this._codeMirror.getScrollInfo();
-        this._codeMirror.setValue(nextProps.value);
-        this._codeMirror.scrollTo(
-          prevScrollPosition.left,
-          prevScrollPosition.top
-        );
-      } else {
-        this._codeMirror.setValue(nextProps.value);
-      }
-    } else if (!nextProps.value) {
-      this._codeMirror.setValue("");
-    }
+    viewRef.current ??= new EditorView({
+      state: editorState,
+      parent: parentRef.current,
+    });
 
-    if (typeof nextProps.options === "object") {
-      for (const optionName in nextProps.options) {
-        if (nextProps.options.hasOwnProperty(optionName)) {
-          this._updateOption(optionName, nextProps.options[optionName]);
-        }
+    return () => {
+      if (viewRef.current) {
+        viewRef.current.destroy();
+        viewRef.current = null;
       }
-    }
-  }
+    };
+  }, []);
 
-  focus() {
-    if (this._codeMirror) {
-      this._codeMirror.focus();
+  // handle value prop updates
+  useEffect(() => {
+    if (value == null) {
+      return;
     }
-  }
-
-  render() {
-    return (
-      <textarea
-        autoComplete="off"
-        autoFocus={this.props.autoFocus}
-        defaultValue={this.props.value}
-        ref={this._setTextAreaRef}
-        placeholder={this.props.placeholder}
-      />
-    );
-  }
-
-  _updateOption(optionName: string, newValue: any) {
-    const oldValue = this._codeMirror.getOption(optionName);
-
-    if (oldValue !== newValue) {
-      this._codeMirror.setOption(optionName, newValue);
+    const currentValue = viewRef.current?.state.doc.toString();
+    if (viewRef.current && value !== currentValue) {
+      viewRef.current.dispatch({
+        changes: {
+          from: 0,
+          to: currentValue.length,
+          insert: value,
+        },
+        effects: [
+          preserveScrollPosition &&
+            EditorView.scrollIntoView(0, {
+              yMargin: -viewRef.current.scrollDOM.scrollTop,
+            }),
+        ].filter(Boolean),
+      });
     }
-  }
-
-  _onChange = (doc: any, change: any) => {
-    if (change.origin !== "setValue") {
-      this.props.onChange(doc.getValue());
+  }, [value]);
+
+  // handle lineWrapping updates
+  useEffect(() => {
+    if (viewRef.current) {
+      // todo: investigate why the dispatch does not work and remove the DOM manipulations below
+      // viewRef.current.dispatch({
+      //   effects: lineWrappingCompartment.reconfigure(
+      //     options.lineWrapping ? EditorView.lineWrapping : []
+      //   ),
+      // });
+      if (options.lineWrapping) {
+        parentRef.current
+          .querySelector(".cm-content")
+          .classList.add("cm-lineWrapping");
+      } else {
+        parentRef.current
+          .querySelector(".cm-content")
+          .classList.remove("cm-lineWrapping");
+      }
```

Comment:
I have to resort to DOM manipulations as the compartment update approach does not work, though it works for simple cases: https://codesandbox.io/p/sandbox/codemirror-6-demo-pl8dc

 Helps are welcomed.